masterkidan/NoSQLProvider
# Add more logs for DB connection failures

## Overview
This PR aims to add more logging around to help address issues where users' IDB connections were being dropped due to unexpected failures (specifically related to file systems).

This can be achieved by tracking `onclose` event fired by the `IDBDatabase` connection object to capture details when the DB connection is closed by the system. This event is based on the [close pending flag](https://w3c.github.io/IndexedDB/#connection-close-pending-flag) which tells us if the connection closure was unexpected.


## Details
The current approach is to make use of the `onclose` event that is fired when "the database is unexpectedly closed. This can happen, for example, when the application is shut down or access to the disk the database is stored on is lost while the database is open." (See [IDBDatabase.onclose Mozilla docs](https://developer.mozilla.org/en-US/docs/Web/API/IDBDatabase/onclose))

`onclose` event payload:
![onclose event payload](https://user-images.githubusercontent.com/17183826/106790480-7bef8b80-6608-11eb-8a89-3bfe9cdbb805.png)


These are the attributes that can be used for logging
```
event.target: {
  name,
  objectStoreNames,
  type
}
```

## Testing
  - Set a breakpoint in `handleOnClose()`
  - Delete databases in IndexedDb (Application tab in Dev Tools to fire the `IDBDatabase.onclose` event
  - Observe console logs for the error message
